src/ripple/app/tx/impl/SetOracle.cpp

```diff
@@ -0,0 +1,284 @@
+*/
+//==============================================================================
+
+#include <ripple/app/tx/impl/SetOracle.h>
+#include <ripple/ledger/Sandbox.h>
+#include <ripple/ledger/View.h>
+#include <ripple/protocol/Feature.h>
+#include <ripple/protocol/Rules.h>
+#include <ripple/protocol/TxFlags.h>
+#include <ripple/protocol/digest.h>
+
+namespace ripple {
+
+static inline uint256
+tokenPairHash(STObject const& pair)
+{
+    return sha512Half(
+        pair.getFieldCurrency(sfBaseAsset).currency(),
+        pair.getFieldCurrency(sfQuoteAsset).currency());
+}
+
+NotTEC
+SetOracle::preflight(PreflightContext const& ctx)
+{
+    if (!ctx.rules.enabled(featurePriceOracle))
+        return temDISABLED;
+
+    if (auto const ret = preflight1(ctx); !isTesSuccess(ret))
+        return ret;
+
+    if (ctx.tx.getFlags() & tfUniversalMask)
+        return temINVALID_FLAG;
+
+    auto const& dataSeries = ctx.tx.getFieldArray(sfPriceDataSeries);
+    if (dataSeries.empty())
+        return temARRAY_EMPTY;
+    if (dataSeries.size() > maxOracleDataSeries)
+        return temARRAY_TOO_LARGE;
+
+    auto isInvalidLength = [&](auto const& sField, std::size_t length) {
+        return ctx.tx.isFieldPresent(sField) &&
+            ctx.tx[sField].length() > length;
+    };
+
+    if (isInvalidLength(sfProvider, maxOracleProvider) ||
+        isInvalidLength(sfURI, maxOracleURI) ||
+        isInvalidLength(sfAssetClass, maxOracleSymbolClass))
+        return temMALFORMED;
+
+    return preflight2(ctx);
+}
+
+TER
+SetOracle::preclaim(PreclaimContext const& ctx)
+{
+    auto const sleSetter =
+        ctx.view.read(keylet::account(ctx.tx.getAccountID(sfAccount)));
+    if (!sleSetter)
+        return terNO_ACCOUNT;
+
+    // lastUpdateTime must be within maxLastUpdateTimeDelta seconds
+    // of the last closed ledger
+    using namespace std::chrono;
+    std::size_t const closeTime =
+        duration_cast<seconds>(ctx.view.info().closeTime.time_since_epoch())
+            .count();
+    std::size_t const lastUpdateTime = ctx.tx[sfLastUpdateTime];
+    if (lastUpdateTime <= epoch_offset.count())
+        return tecINVALID_UPDATE_TIME;
+    std::size_t const lastUpdateTimeEpoch =
+        lastUpdateTime - epoch_offset.count();
+    if (lastUpdateTimeEpoch < closeTime ||
+        lastUpdateTimeEpoch > (closeTime + maxLastUpdateTimeDelta))
+        return tecINVALID_UPDATE_TIME;
+
+    auto const sle = ctx.view.read(keylet::oracle(
+        ctx.tx.getAccountID(sfAccount), ctx.tx[sfOracleDocumentID]));
+
+    hash_set<uint256> pairs;
+    hash_set<uint256> pairsDel;
+    for (auto const& entry : ctx.tx.getFieldArray(sfPriceDataSeries))
+    {
+        auto const hash = tokenPairHash(entry);
+        if (pairs.contains(hash))
+            return tecDUPLICATE;
+        if (entry.isFieldPresent(sfAssetPrice))
+            pairs.emplace(hash);
+        else if (sle)
+            pairsDel.emplace(hash);
+        else
+            return temMALFORMED;
+    }
```

Comment:
Why is this not in `preflight`? Seems more efficient that way. Line 104 might be the only exception, but that processing could just be checked in `doApply`.

Reply:
We need to figure out the reserve requirement in `preclaim`. And we have to iterate trough the array to figure it out.